### PR TITLE
[EVM] subscription based eth endpoints

### PR DESCRIPTION
**Pull request by jewei1997**
## Describe your changes and provide context
Implementations of eth_subscribe for newHeads and logs. newPendingTransactions is not supported because it is hard to get transactions from tendermint's mempool.

## Testing performed to validate your change
unit tests + local testing for newHeads

### CODE REVIEW

File: evmrpc/subscribe.go [hunk 1]
```diff
@@ -4,138 +4,209 @@ import (
...
-func NewBlockQueryBuilder() *QueryBuilder {
-	return &QueryBuilder{
-		conditions: []string{
-			"tm.event = 'NewBlock'",
-		},
-	}
-}
+	switch eventName {
+	case "newHeads":
+		subscriberID, subCh, err := a.subscriptionManager.Subscribe(ctx, NewHeadQueryBuilder(), 100)
```
Comment (codchen) [inline evmrpc/subscribe.go]: let's make the limi a config instead of hardcoding

File: evmrpc/subscribe.go [hunk 2]
```diff
@@ -4,138 +4,209 @@ import (
...
+			return nil, err
+		}
+
+		go func() {
+			for {
+				select {
+				case res := <-subCh:
+					ethHeader, err := encodeTmHeader(res.Data.(*tmtypes.EventDataNewBlockHeader))
+					if err != nil {
+						_ = a.subscriptionManager.Unsubscribe(ctx, subscriberID)
```
Comment (codchen) [inline evmrpc/subscribe.go]: nit: can just do `defer _ = a.subscriptionManager.Unsubscribe(ctx, subscriberID)` immediately after `go func(){`

File: evmrpc/subscribe.go [hunk 3]
```diff
@@ -4,138 +4,209 @@ import (
...
+					return
+				case <-notifier.Closed():
+					_ = a.subscriptionManager.Unsubscribe(ctx, subscriberID)
+					return
+				}
+			}
+		}()
+		return rpcSub, nil
+	case "logs":
+		resultEventAllAddrs := make(chan coretypes.ResultEvent)
```
Comment (codchen) [inline evmrpc/subscribe.go]: is it intentional to use unbuffered channel here?

Comment (jewei1997) [reply evmrpc/subscribe.go]: updated to buffered channel of sufficient capacity.

File: evmrpc/subscribe.go (hunk 2 above)
Comment (jewei1997) [reply evmrpc/subscribe.go]: updated.

File: evmrpc/subscribe.go (hunk 1 above)
Comment (jewei1997) [reply evmrpc/subscribe.go]: updated.

**Pull request merged**